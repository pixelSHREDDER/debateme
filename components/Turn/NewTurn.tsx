'use client'

import addTurn from '@/actions/add-turn'
import { TDebate } from '@/lib/prisma-types'
import { useFormState } from 'react-dom'
import { useRef, useState } from 'react'
import { useUser } from '@auth0/nextjs-auth0/client'
import Editor from '@/components/Editor/Editor'
import FormSubmitButton from '@/components/Forms/FormSubmitButton'
import FormAlert from '@/components/Forms/FormAlert'
import { Button, Fieldset, Flex, Modal, Stack, Text } from '@mantine/core'
import Quotes from '@/components/Quotes/Quotes'
import { useDisclosure } from '@mantine/hooks'
import useCustomMarks from '@/hooks/useCustomMarks'

const SUBMIT_WARNING = 'Remember: turns can\'t be edited, and you can\'t submit another turn until your opponent finishes theirs.'

const initialState = {
  message: '',
}

export default function NewTurn({ debate, onSubmit }: { debate: TDebate, onSubmit: Function }) {
  const [isProofread, setIsProofread] = useState<boolean>(false)
  const [state, formAction] = useFormState(addTurn, initialState)
  const bodyRef = useRef<HTMLInputElement>(null)
  const { user } = useUser()
  const [opened, { open, close }] = useDisclosure(false)
  const { stripCustomMarks } = useCustomMarks()

  const onUpdate = (editorHtml?: string) => {
    if (!!editorHtml && bodyRef.current) bodyRef.current.value = stripCustomMarks(editorHtml)
  }

  const onProofread = () => { if (!isProofread) setIsProofread(true) }

  const onTurnSubmit = (payload: FormData) => {
    formAction(payload)
    if (onSubmit) {
      onSubmit()
    }
  }

  if (!user || !user.sub) {
    return ''
  }

  return (
    <form action={onTurnSubmit} id="new-turn-form" data-testid="new-turn-form">
      <Fieldset mt={30} mx={0} radius="lg" variant="filled" legend="It's your turn!">
        <input
          type="hidden"
          id="debateId"
          name="debateId"
          value={debate.id} />
        <input
          type="hidden"
          id="isCreator"
          name="isCreator"
          value={user.sub === debate.creatorSub ? 'true' : 'false'} />
        <input
          type="hidden"
          id="userSub"
          name="userSub"
          value={user.sub} />
        <input
          ref={bodyRef}
          type="hidden"
          id="bodyString"
          name="bodyString" />
        {/*<InputLabel htmlFor="body">New turn text</InputLabel>*/}
        <Editor
          id="body"
          onProofread={onProofread}
          onUpdate={onUpdate}
          required="true"
          testid="new-turn-editor">
        </Editor>
        {state?.message &&
          <FormAlert
            message={state.message}
            title="Something went wrong"
            type={2} />
        }
        <Flex align="center" gap={20} justify="flex-start">
          <Quotes />
          {!!isProofread &&
            <Button flex="1 0 auto" onClick={open}>Finish Your Turn</Button>
          }
        </Flex>
      </Fieldset>
      <Modal opened={opened} onClose={close} centered title="All set?">
        <Stack>
          <Text>{SUBMIT_WARNING}</Text>
          <Flex gap="sm">
            <FormSubmitButton form="new-turn-form" label="I'm All Done" />
            <Button variant="outline" onClick={close}>Not Yet</Button>
          </Flex>
        </Stack>
      </Modal>
    </form>
  )
}
