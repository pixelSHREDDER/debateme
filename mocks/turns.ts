import { TTurn } from '@/lib/prisma-types'

// Lorem Ipsum HTML courtesy of https://css-tricks.com/new-site-html-ipsum/

export const createTurns: () => TTurn[] = () => {
  const createdAt = new Date('2024-04-13T12:24:02Z')

  return [
    {
      id: 1,
      body: `
          <p>Didn't melt fairer keepsakes since Fellowship elsewhere.</p>
          <p>Woodlands payment <a href="#">Osgiliath</a> tightening. Barad-dur follow belly comforts tender tough bell? Many that live deserve death. Some that die deserve life. Outwitted teatime grasp defeated before stones reflection corset seen animals Saruman's call?</p>
          <p>Curse you and all the halflings! Deserted anytime Lake-town burned caves balls. Smoked lthilien forbids Thrain? <span style="font-size: 2.0rem">Tad survive ensnare joy mistake courtesy Bagshot Row.</span> Ligulas step drops both? You shall not pass!</p>
          <p>Branch hey-diddle-diddle pony trouble'll sleeping during jump Narsil. <span style="font-size: 0.6rem">North valor overflowing sort Iáve mister kingly money?</span> Tender respectable success Valar impressive unfriendly bloom scraped?</p>
          <ul>
            <li>Adamant.</li>
            <li>Southfarthing!</li>
            <li>Witch-king.</li>
            <li>Precious.</li>
            <li>Gaffer's!</li>
          </ul>
          <ul>
            <li>Excuse tightening yet survives two cover Undómiel city ablaze.</li>
            <li>Keepsakes deeper clouds Buckland position 21 lied bicker fountains ashamed.</li>
            <li>Women rippling cold steps rules Thengel finer.</li>
            <li>Portents close Havens endured irons hundreds handle refused sister?</li>
            <li>Harbor Grubbs fellas riddles afar!</li>
          </ul>
          <p>Ravens wonder wanted runs me crawl gaining lots faster! Khazad-dum surprise baby season ranks. I bid you all a very fond farewell.</p>
          <ol>
            <li>Narsil.</li>
            <li>Elros.</li>
            <li>Arwen Evenstar.</li>
            <li>Maggot's?</li>
            <li>Bagginses?</li>
          </ol>
          <ol>
            <li>Concerning Hobbits l golf air fifth bell prolonging camp.</li>
            <li>Grond humble rods nearest mangler.</li>
            <li>Enormity Lórien merry gravy stayed move.</li>
            <li>Diversion almost notion furs between fierce laboring Nazgûl ceaselessly parent.</li>
            <li>Agree ruling um wasteland Bagshot Row expect sleep.</li>
          </ol>
          <p>Delay freezes Gollum. Let the Ring-bearer decide. Bagshot Row chokes pole pauses immediately orders taught éored musing three-day? Disease rune repel source fire Goblinses already?</p>
          <hr>
          <p>What about second breakfast? Nags runt near Lindir lock discover level? Andûril breathe waited flatten union.</p>
          <blockquote>
            <p>You shall be the Fellowship of the Ring.</p>
            <footer>—Númenor, <cite>sweeter burned verse</cite></footer>
          </blockquote>
          <p>I think we should get off the road. Penalty sight splintered Misty Mountain mithril? Unrest lasts rode league bears absence Bracegirdle athletic contract nice parent slowed?</p>
          <pre>Pardon Concerning Hobbits rune goblins? Twitching figure including rightful Thorin's level! Worth tubers threats Hornburg deadliest? Unfold thumping shh wants Homely!</pre>
          <hr>`,
      debateId: 1,
      createdAt,
      userSub: 'auth0|1234567890abcdefghij1234',
      },
      {
        id: 2,
        body: `
          <p>Didn't melt fairer keepsakes since Fellowship elsewhere.</p>
          <p>Woodlands payment <a href="#">Osgiliath</a> tightening. Barad-dur follow belly comforts tender tough bell? Many that live deserve death. Some that die deserve life. Outwitted teatime grasp defeated before stones reflection corset seen animals Saruman's call?</p>
          <p><span style="font-size: 2.0rem">Tad survive ensnare joy mistake courtesy Bagshot Row.</span></p>
          <p>Ligulas step drops both? You shall not pass! Tender respectable success Valar impressive unfriendly bloom scraped? Branch hey-diddle-diddle pony trouble'll sleeping during jump Narsil.</p>
          <p><span style="font-size: 0.6rem">North valor overflowing sort Iáve mister kingly money?</span></p>
          <p>Curse you and all the halflings! Deserted anytime Lake-town burned caves balls. Smoked lthilien forbids Thrain?</p>
          <ul>
            <li>Adamant.</li>
            <li>Southfarthing!</li>
            <li>Witch-king.</li>
            <li>Precious.</li>
            <li>Gaffer's!</li>
          </ul>
          <ul>
            <li>Excuse tightening yet survives two cover Undómiel city ablaze.</li>
            <li>Keepsakes deeper clouds Buckland position 21 lied bicker fountains ashamed.</li>
            <li>Women rippling cold steps rules Thengel finer.</li>
            <li>Portents close Havens endured irons hundreds handle refused sister?</li>
            <li>Harbor Grubbs fellas riddles afar!</li>
          </ul>
          <p>Ravens wonder wanted runs me crawl gaining lots faster! Khazad-dum surprise baby season ranks. I bid you all a very fond farewell.</p>
          <ol>
            <li>Narsil.</li>
            <li>Elros.</li>
            <li>Arwen Evenstar.</li>
            <li>Maggot's?</li>
            <li>Bagginses?</li>
          </ol>
          <ol>
            <li>Concerning Hobbits l golf air fifth bell prolonging camp.</li>
            <li>Grond humble rods nearest mangler.</li>
            <li>Enormity Lórien merry gravy stayed move.</li>
            <li>Diversion almost notion furs between fierce laboring Nazgûl ceaselessly parent.</li>
            <li>Agree ruling um wasteland Bagshot Row expect sleep.</li>
          </ol>
          <p>Delay freezes Gollum. Let the Ring-bearer decide. Bagshot Row chokes pole pauses immediately orders taught éored musing three-day? Disease rune repel source fire Goblinses already?</p>
          <hr>
          <p>What about second breakfast? Nags runt near Lindir lock discover level? Andûril breathe waited flatten union.</p>
          <blockquote>
            <p>You shall be the Fellowship of the Ring.</p>
            <footer>—Númenor, <cite>sweeter burned verse</cite></footer>
          </blockquote>
          <p>I think we should get off the road. Penalty sight splintered Misty Mountain mithril? Unrest lasts rode league bears absence Bracegirdle athletic contract nice parent slowed?</p>
          <pre>Pardon Concerning Hobbits rune goblins? Twitching figure including rightful Thorin's level! Worth tubers threats Hornburg deadliest? Unfold thumping shh wants Homely!</pre>`,
        debateId: 1,
        createdAt,
        userSub: 'auth0|abcdefghij1234567890abcd',
        },
    ] as const
  }
