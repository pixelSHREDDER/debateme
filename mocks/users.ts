import { UserProfile } from '@auth0/nextjs-auth0/client'

export const getUsers: () => UserProfile[] = () => [
    {
      nickname: 'ron.swanson12345',
      name: 'Ron Swanson',
      picture: 'https://s.gravatar.com/avatar/7642279092c16299a6a93e7523cc9ed2?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Frs.png',
      updated_at: '2024-08-17T04:12:58.145Z',
      email: 'ron.swanson12345@excellent-parser.name',
      email_verified: false,
      sub: 'auth0|1234567890abcdefghij1234',
      sid: 'o_abcdefgh12345678-ABCDEF12345',
    },
    {
      nickname: 'leslie.knope54321',
      name: 'Leslie Knope',
      picture: 'https://s.gravatar.com/avatar/7642279092c16299a6a93e7523cc9ed2?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Flk.png',
      updated_at: '2024-08-17T04:12:58.145Z',
      email: 'leslie.knope54321@excellent-parser.name',
      email_verified: false,
      sub: 'auth0|abcdefghij1234567890abcd',
      sid: 'o_12345678abcdefgh-12345ABCDEF',
    },
  ] as const
