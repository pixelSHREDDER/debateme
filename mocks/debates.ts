import { TDebate } from '@/lib/prisma-types'
import { DebateStatus } from '@prisma/client'
import { createTurns } from './turns'

export const createDebates: () => TDebate[] = () => {
  const createdAtDate = new Date('2024-04-13T12:24:02Z')
  const cooldownDate = new Date()
  const createdTurns = createTurns()

  return [
    {
      id: 1,
      topic: 'Your turn',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      turn: createdTurns,
      status: DebateStatus.CreatorTurn,
    },
    {
      id: 2,
      topic: 'Opponent\'s Turn',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      status: DebateStatus.OpponentTurn,
      turn: [
        createdTurns[0],
      ],
    },
    {
      id: 3,
      topic: 'Wait for Cooldown',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      status: DebateStatus.CreatorCooldown,
      turn: [
        createdTurns[0],
        {
          ...createdTurns[1],
          createdAt: cooldownDate,
        },
      ],
    },
    {
      id: 4,
      topic: 'No opponent',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: null,
      cooldownMins: 60,
      status: DebateStatus.NoOpponent,
    },
    {
      id: 5,
      topic: 'You\'re Invited to join',
      createdAt: createdAtDate,
      creatorSub: 'auth0|abcdefghij1234567890abcd',
      opponentSub: null,
      cooldownMins: 60,
      status: DebateStatus.NoOpponent,
    },
  ] as const
}
