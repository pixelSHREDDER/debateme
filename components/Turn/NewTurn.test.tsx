import * as auth0Hooks from '@auth0/nextjs-auth0/client'
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import ProviderWrapper from '@/tests/helpers/ProviderWrapper'
import { getUsers } from '@/mocks/users'
//import {userEvent} from '@testing-library/user-event'
import { createDebates } from '@/mocks/debates'
import NewTurn from './NewTurn'

const debates = createDebates()
const users: auth0Hooks.UserProfile[] = getUsers()

// TODO: Circle back on mocking Editor
/*vi.mock('@/components/Editor/Editor', () => ({
  default: vi.fn((onUpdate: ChangeEventHandler<HTMLInputElement>) => <input type="text" data-testid="editor-input" onChange={onUpdate} />),
}))*/

describe('New Turn component', () => {
  //const useFormStateSpy = vi.spyOn(reactDom, 'useFormState')
  const useUserSpy = vi.spyOn(auth0Hooks, 'useUser')

  beforeEach(() => {
    vi.useFakeTimers()
    useUserSpy.mockReturnValue({
      checkSession: vi.fn(),
      isLoading: false,
      user: users[0],
    })
    //useFormStateSpy.mockImplementation(() => ([{}, vi.fn(), false]))
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  describe('New Turn', () => {
    test('renders correctly', () => {
      render(
        <NewTurn debate={debates[0]} onSubmit={vi.fn()} />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.getByTestId('new-turn-form')).toMatchSnapshot()
    })

    test('does not render if user is loading', () => {
      useUserSpy.mockReturnValueOnce({
        checkSession: vi.fn(),
        isLoading: true,
        user: users[0],
      })
      render(
        <NewTurn debate={debates[0]} onSubmit={vi.fn()} />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.queryByTestId('new-turn-form')).toBeNull
    })

    test('does not render if user is not defined', () => {
      useUserSpy.mockReturnValueOnce({
        checkSession: vi.fn(),
        isLoading: false,
      })
      render(
        <NewTurn debate={debates[0]} onSubmit={vi.fn()} />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.queryByTestId('new-turn-form')).toBeNull
    })
  })

  describe('Submitting new turn', () => {
    test('submits new turn when button is clicked', async () => {
      //const user = userEvent.setup()
      const onSubmit = vi.fn()
      render(
        <NewTurn debate={debates[0]} onSubmit={onSubmit} />,
        { wrapper: ProviderWrapper },
      )
      //const submit = screen.getByTestId('derp')
      //fireEvent.click(submit)
      //await user.click(submit)
      fireEvent.submit(screen.getByTestId('new-turn-form'))
      //act(() => vi.advanceTimersByTime(1000))
      //await user.click(screen.getByTestId('derp'))

      //expect(onSubmit).toHaveBeenCalled()
    })
  })

  /*describe('Editor updates', () => {
    test('updates body ref when Editor contents are changed', async () => {
      render(
        <NewTurn debate={debates[0]} onSubmit={vi.fn()} />,
        { wrapper: ProviderWrapper },
      )
      const user = userEvent.setup()
      const editor = await screen.findByTestId('editor-input')

      expect(editor).not.toBeNull

      await user.type(editor, 'New content')
      // eslint-disable-next-line testing-library/no-node-access
      *//*fireEvent.change(editor.getElementsByTagName('p')[0], {
        target: { textContent: 'New content' },
      })*//*

      expect(editor).toMatchInlineSnapshot(`
          <p>
            New content
          </p>
      `)

        //expect(mockOnChange).toHaveBeenCalledTimes(11)
        //expect(mockOnChange).toHaveBeenCalledWith('<p>New content</p>')
    })
  })*/
})
