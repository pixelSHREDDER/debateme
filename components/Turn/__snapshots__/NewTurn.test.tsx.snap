// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`New Turn component > New Turn > renders correctly 1`] = `
<form
  data-testid="new-turn-form"
>
  <fieldset
    class="m_ef274e49 m_e9408a47 mantine-Fieldset-root"
    data-variant="filled"
    radius="lg"
    style="--fieldset-radius: var(--mantine-radius-lg); margin-inline: 0rem;"
  >
    <legend
      class="m_90794832 mantine-Fieldset-legend"
    >
      It's your turn!
    </legend>
    <input
      id="debateId"
      name="debateId"
      type="hidden"
      value="1"
    />
    <input
      id="isCreator"
      name="isCreator"
      type="hidden"
      value="1"
    />
    <input
      id="userSub"
      name="userSub"
      type="hidden"
      value="auth0|1234567890abcdefghij1234"
    />
    <input
      id="bodyString"
      name="bodyString"
      type="hidden"
    />
    <div
      class="m_8bffd616 mantine-Flex-root __m__-r1"
      style="gap: calc(1.25rem * var(--mantine-scale)); align-items: center; justify-content: flex-start;"
    >
      <aside
        aria-describedby="alert_quotes_:r3:-body"
        class="_quotes_e6219c m_66836ed3 mantine-Alert-root"
        data-variant="outline"
        id="alert_quotes_:r3:"
        role="alert"
        style="--alert-radius: var(--mantine-radius-md); --alert-bg: transparent; --alert-color: #9562d2; --alert-bd: calc(0.0625rem * var(--mantine-scale)) solid #9562d2; font-style: italic; width: auto; max-width: 75%;"
      >
        <div
          class="m_a5d60502 mantine-Alert-wrapper"
        >
          <div
            class="m_667c2793 mantine-Alert-body"
          >
            <div
              class="m_7fa78076 mantine-Alert-message"
              data-variant="outline"
              id="alert_quotes_:r3:-body"
            >
              “Don’t be a dick.” ― Unknown
            </div>
          </div>
        </div>
      </aside>
      <button
        class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
        data-variant="filled"
        style="--button-radius: var(--mantine-radius-md); --button-bg: var(--mantine-color-marble-filled); --button-hover: var(--mantine-color-marble-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; flex-shrink: 0;"
        type="submit"
      >
        <span
          class="m_80f1301b mantine-Button-inner"
        >
          <span
            class="m_811560b9 mantine-Button-label"
          >
            Finish Your Turn
          </span>
        </span>
      </button>
    </div>
  </fieldset>
</form>
`;
